File: import_bloom.py
from blpapi import SessionOptions, Session, SubscriptionList, Event, Name #type: ignore

def handle_event(event, session):
    et = event.eventType()

    for msg in event:
        if et == Event.SESSION_STATUS:
            print("[SESSION]", msg.messageType(), msg.toString())

            if msg.messageType() == Name("SessionStarted"):
                print("Session started → subscribing")

                subs = SubscriptionList()
                subs.add("IBM US Equity",  "BID,ASK,LAST_PRICE")
                subs.add("MSFT US Equity", "BID,ASK,LAST_PRICE")

                session.subscribe(subs)

            elif msg.messageType() in (
                Name("SessionTerminated"),
                Name("SessionStartupFailure"),
            ):
                print("Session failed or terminated")
                session.stop()

        elif et == Event.SUBSCRIPTION_DATA:
            cid = msg.correlationId().value()
            if msg.hasElement("LAST_PRICE"):
                last = msg.getElement("LAST_PRICE").getValueAsFloat()
                print(f"[DATA] cid={cid} LAST_PRICE={last}")

        elif et == Event.SUBSCRIPTION_STATUS:
            print("[SUB_STATUS]", msg.toString())


def main():
    opts = SessionOptions()
    opts.setServerHost("localhost")
    opts.setServerPort(8194)

    session = Session(opts, eventHandler=handle_event)
    session.startAsync()

    input("Press ENTER to quit\n")
    session.stop()


if __name__ == "__main__":
    main()

File: import_bloom.cpp
#include <iostream>
#include <blpapi_event.h>
#include <blpapi_eventhandler.h>
#include <blpapi_message.h>
#include <blpapi_messageiterator.h>
#include <blpapi_name.h>
#include <blpapi_session.h>
#include <blpapi_sessionoptions.h>
#include <blpapi_subscriptionlist.h>
#include <blpapi_correlationid.h>
using namespace blpapi;



class SubscriptionEventHandler : public EventHandler {
public:
    void processEvent(const Event& event, Session* session) override {
        // Log event type
        std::cout << "== EventType: " << event.eventType() << std::endl;

        MessageIterator it(event);
        while (it.next()) {
            Message msg = it.message();
            std::cout << "  msgType=" << msg.messageType()
                      << "  cid=" << msg.correlationId() << std::endl;

            if (event.eventType() == Event::SUBSCRIPTION_DATA) {
                if (msg.hasElement("LAST_PRICE")) {
                    double last = msg.getElementAsFloat64("LAST_PRICE");
                    std::cout << "    LAST_PRICE=" << last << std::endl;
                }
            } else if (event.eventType() == Event::SUBSCRIPTION_STATUS) {
                // Expect SubscriptionStarted / SubscriptionFailure / SubscriptionTerminated
                msg.print(std::cout);
            } else if (event.eventType() == Event::SESSION_STATUS) {
                // Expect SessionStarted / SessionTerminated / SessionStartupFailure
                Name terminated("SessionTerminated");
                Name startupFail("SessionStartupFailure");
                if (msg.messageType() == terminated || msg.messageType() == startupFail) {
                    std::cout << "Session ending; calling stop()." << std::endl;
                    if (session) session->stop();
                }
            }
        }
    }
};

int main() {
    // Session options
    SessionOptions opts;
    opts.setServerHost("localhost");
    opts.setServerPort(8194);

    // Handler + async session
    SubscriptionEventHandler handler;
    Session session(opts, &handler);

    if (!session.startAsync()) {
        std::cerr << "Failed to start async session." << std::endl;
        return 1;
    }

    // Optionally open service explicitly; subscribe will also trigger it if needed
    // session.openServiceAsync("//blp/mktdata");

    // Build and submit subscriptions (after SessionStarted arrives)
    // For brevity submit immediately; in production, gate on SessionStarted
    SubscriptionList subs;
    subs.add("IBM US Equity",  "BID,ASK,LAST_PRICE", "", CorrelationId(1));
    subs.add("MSFT US Equity", "BID,ASK,LAST_PRICE", "", CorrelationId(2));
    session.subscribe(subs);

    std::cout << "Press ENTER to quit" << std::endl;
    (void)getchar();

    session.stop();  // Ensure clean shutdown
    return 0;
}
